text/0000-scoped-take-2.md

```diff
@@ -0,0 +1,262 @@
+
+On a separate note, in practice one will usually not want to spawn full-fledged
+threads when doing data-parallel, fork-join style computations; instead you want
+to use a thread pool, work stealing, and higher-level combinators. One nice
+aspect of `Scope` is that crates providing such parallelism frameworks can
+easily hook into the API with their own means of spawning lightweight tasks;
+this should also facilitate a [decoupling of thread pools from the lifetimes
+bounding the tasks they are running](https://github.com/rust-lang/threadpool/issues/7).
+
+## Implementation
+
+Here's a sketch of the implementation, inspired in part by
+[@arielb1's ideas](https://gist.github.com/arielb1/5eb299a87546ce8829b3):
+
+```rust
+pub struct Scope<'a> {
+    dtors: RefCell<Option<DtorChain<'a>>>
+}
+
+struct DtorChain<'a> {
+    dtor: Box<FnBox() + 'a>,
+    next: Option<Box<DtorChain<'a>>>
+}
+
+pub fn scope<'a, F, R>(f: F) -> R where F: FnOnce(&Scope<'a>) -> R {
+    let mut scope = Scope { dtors: RefCell::new(None) };
+    let ret = f(&scope);
+    scope.drop_all();
+    ret
+}
+
+impl<'a> Scope<'a> {
+    // This method is carefully written in a transactional style, so
+    // that it can be called directly and, if any dtor panics, can be
+    // resumed in the unwinding this causes. By initially running the
+    // method outside of any destructor, we avoid any leakage problems
+    // due to #14875.
+    fn drop_all(&mut self) {
+        loop {
+            // use a separate scope to ensure that the RefCell borrow
+            // is relinquished before running `dtor`
+            let dtor = {
+                let mut dtors = self.dtors.borrow_mut();
+                if let Some(mut node) = dtors.take() {
+                    *dtors = node.next.take().map(|b| *b);
+                    node.dtor
+                } else {
+                    return
+                }
+            };
+            dtor()
+        }
+    }
+
+    pub fn defer<F>(&self, f: F) where F: FnOnce() + 'a {
+        let mut dtors = self.dtors.borrow_mut();
+        *dtors = Some(DtorChain {
+            dtor: Box::new(f),
+            next: dtors.take().map(Box::new)
+        });
+    }
+}
+
+impl<'a> Drop for Scope<'a> {
+    fn drop(&mut self) {
+        self.drop_all()
+    }
+}
+```
+
+This implementation does a few interesting things:
+
+* It avoids any allocation in the case of a single deferred computation.
+
+* It works around issue 14875 in a somewhat subtle way: the `drop_all` method is
+  called *both* normally within `scope`, and in the `Scope` destructor. The
+  method is also coded transactionally. This means that the first panic (if any)
+  in a deferred computation triggers the drop for `Scoped` (without any leakage,
+  avoiding #14875), and any remaining panics wind up aborting.
+
+Note that this is just a sketch: the use of a linked list here, and `RefCell`
+internally, could both be switched for something more interesting.
+
+# Drawbacks
```

Comment:
More than the ergonomics, the fundamental drawback of this approach is that it is an inversion of the normal RAII workflow: functions acquiring some resource return a guard (representing the resource itself) which can be dropped to release the resource. If you acquire an RAII resource internally, you can provide the same API by forwarding the guard.

With APIs like the above, this is reversed - to use an RAII resource, you accept a guard as an argument rather than returning it, and you cannot acquire the resource internally and return it, you *must* have it passed to you.

This is not necessarily a death blow, just a significant drawback in terms of consistency. For instance, the original `thread::scoped` API meant that `thread::spawn` was really just a special case of `scoped` where `'a` is `'static` and the default behavior of the guard is different. This API is fundamentally different, and encourages totally different patterns.